### ISSUE TITLE
TypeScript compile issues with VFS files. 

### ISSUE THREAD

**Issue by MarketingPip** [sebastianwessel/quickjs#47]
Not really sure what is causing the error. But appears to be with strings etc stored in fs.. (maybe improper usage of my file system)?

I do believe it has something to do with `math.json` and `test.json` file - being invalid JSON and typescript not being able to catch it. Tho not sure why it is touching it at all. (If my assumptions are correct). 

If right tho - could this not cause tons of problems with files non js based in file system...?

To reproduce try using this file system: 

```json
{
    "math.json": "{\n  \"func\": \"async function(args){ const math = (await import('https://esm.sh/mathjs')); const result = math.evaluate(args.join(' ')); return result ;}\",\n  \"usage\": \"\"\n}",
    "test.json": "{\n  \"func\": \"async args => { term.echo('hello') }\",\n  \"help\": \"Usage: says hello\"\n}",
    "test.cpp": "// C++ program to display \"Hello World\"\n  \n// Header file for input output functions\n#include <iostream>\nusing namespace std;\n  \n// Main() function: where the execution of program begins\nint main()\n{\n    // prints hello world\n    cout << \"Hello World\";\n  \n    return 0;\n}",
    "test.ts": "function sayHello(name: string): string {\n    return `Hello, ${name}!`;\n}\n\nconst message = sayHello('World');\nconsole.log(message);\n",
    "hello.php": "<?php\necho \"Hello, World!\";\n?>\n",
    "hello.js": "console.log(\"Hello, World!\");",
    "hello.coffee": "console.log 'Hello, World!'",
    "ansi.js": "const redText = '\u001b[31mThis text is red\u001b[0m';\nconst greenText = '\u001b[32mThis text is green\u001b[0m';\n\n// Log colored text\nconsole.log(redText);\nconsole.log(greenText);",
    "custom.js": "export const relativeImportFunction = ()=>'Hello from relative import function'",
    "main.js": "\nimport { relativeImportFunction } from './custom.js';\n\nconsole.log(relativeImportFunction());\n",
    "textFile.txt": "Hello from the text file!",
    "main2.js": "\nimport { readFileSync, readFile, writeFileSync } from \"https://esm.sh/node:fs\"\nasync function run(){\nreadFile('./src/textFile.txt', 'utf8', (err, data) => {\n  if (err) throw err;\n  console.log(data);\n});\n\nconsole.log(readFileSync('./src/textFile.txt'))\n}\nawait run()\n",
    "fetch.js": "async function getRepoData(owner, repo) {\n  const url = `https://api.github.com/repos/${owner}/${repo}`;\n  \n  try {\n    const response = await fetch(url);\n\n    if (!response.ok) {\n      throw new Error(`GitHub API error: ${response.statusText}`);\n    }\n\n    const data = await response.json();\n    console.log('Repository Data:', data);\n  } catch (error) {\n    console.error(error.message);\n  }\n}\n\nawait getRepoData('octocat', 'Hello-World'); // Replace with the desired owner and repo",
    "fetch_placeholders.js": "async function getJsonPlaceholderData(resource, id) {\n  const url = `https://jsonplaceholder.typicode.com/${resource}/${id}`;\n  \n  try {\n    const response = await fetch(url);\n\n    if (!response.ok) {\n      throw new Error(`API error: ${response.statusText}`);\n    }\n\n    const data = await response.json();\n    console.log('${resource} Data:', data);\n  } catch (error) {\n    console.error(error.message);\n  }\n}\n\nawait getJsonPlaceholderData('posts', 1); // Replace 'posts' and '1' with the desired resource and ID",
    "": {},
    "etc": {
        "motd": "Welcome to the Unix-like virtual filesystem!"
    },
    "var": {
        "log": {
            "syslog": "System log initialized."
        }
    },
    "home": {
        "user": {
            "README.txt": "This is the home directory of the user."
        }
    }
}
```

Edited:

Confirmed - "removed" the two files (in my emulated shell) and when gone from VFS command runs fine....

**Comment by sebastianwessel** [comment]
Can you maybe explain what you try to do and paste the commands & errors?
Otherwise it's a bit hard to follow the issue

**Comment by MarketingPip** [comment]
@sebastianwessel - insert this into the runtime / vfs and you will the issue. 

```json
{
    "math.json": "{\n  \"func\": \"async function(args){ const math = (await import('https://esm.sh/mathjs')); const result = math.evaluate(args.join(' ')); return result ;}\",\n  \"usage\": \"\"\n}",
    "test.json": "{\n  \"func\": \"async args => { term.echo('hello') }\",\n  \"help\": \"Usage: says hello\"\n}"
}
```

It fails to compile due to these "strings".

**Comment by MarketingPip** [comment]
@sebastianwessel - did some digging into this from user side etc... So those JSON files seem to be culprit being they have new lines etc - ie were stringified. 

If you "rename" the files - things will continue to work fine. > meaning somewhere these are trying to get parsed and failing due to new lines / bad escape etc..


I'm curious tho - when invalid json (non stringified) etc in the file system exists. What will the runtime do to handle this? As it should compile...


One option I think should solve this is doing a JSON parse on the files - (not sure if your run time needs it as a JS object or if dirty JSON etc will still work when JSON.parse only returns a string for unescaped JSON.)

tdlr; if you have invalid JSON files currently in your file system (ie containing new lines / un escaped etc). You will get a message saying failure to generate build etc...

**Comment by sebastianwessel** [comment]
Closing this issue, as it is not part of the lib to "fix" wrong inputs.
In real world, the input for the vfs is generated in most cases and not done manually.
Something like:

```ts
{
  'my-file': fs.readFileSync('host/myfile')
}
```

If you do manually stuff, you are also responsiible for correctness

**Comment by MarketingPip** [comment]
@sebastianwessel - I don't think you realize.. This is not BAD inputs etc... lol what if a file gets written to VFS like that... With raw strings etc.. (It should NOT matter how it in the FS). If you did this in a real runtime on your device it would be handled in the runtime with an error based on file type etc... In this case a bad json parse error.... 


> This is ultimately due to TypeScript - not being able to parse these dirty json when getting built.... (What other files mess things up and string - I am not sure...)

**Comment by sebastianwessel** [comment]
> what if a file gets written to VFS like that
It is an in memory file system.

So you have the startup part - where you need to (de-)serialize from host to vfs , and the host is responsible to provide correct input values.

You have the running phase - where the QuickJS is reading/writing to the in memory vfs (like setting/getting vars). Here, the client system is responsible to write stuff, which can be read by itself. It is "just values"

And you might have the final phase, where you might want to serialize the vfs memory to some persistence layer, to be re-used in the next startup part.

So, it is up to the developer, how he is doing the (de-)serialization of the in-memory part. It is not part of this lib, to correct invalid inputs.
In this lib, JSON is the default one, and it is your responsibility as developer, to ensure valid JSON/source code.
If you provide some corrupt values, t is expected to "explode" and not to run quietly.
It is a real bad pattern, change/rewrite the input and to swallow issues from wrong input.

You might want to use JSON for serialization, but you are not restricted to it at all.

See: [https://www.npmjs.com/package/memfs](https://www.npmjs.com/package/memfs)

**Comment by MarketingPip** [comment]
@sebastianwessel - correct I fully understand that. 


Tho again.... due to typescript being compiled- for some reason theses strings need to be parsed as object or dealt with somehow..... 

As if user writes into a vfs BAD JSON it will cause things not to work. Hence my point. 

If "bad JSON" file was in your computer and you ran node - would node still run? Correct. 

(If this is just a TypeScript issue requiring the "vfs" to have parsed JSON objects) possibly mention it in documentation.

**Comment by sebastianwessel** [comment]
You can play around with this:

```ts
import { quickJS } from '@sebastianwessel/quickjs'

const { createRuntime } = await quickJS()

const { evalCode } = await createRuntime({
	allowFs: true, // mount a virtual file system and provide node:fs module
	mountFs: {
		'math.json':
			'CORRUPT THE JSON {\n  "func": "async function(args){ const math = (await import(\'https://esm.sh/mathjs\')); const result = math.evaluate(args.join(\' \')); return result ;}",\n  "usage": ""\n}',
		'test.json': '{\n  "func": "async args => { term.echo(\'hello\') }",\n  "help": "Usage: says hello"\n}',
	},
})

const result = await evalCode(`
import { join } from 'path'
import {readFileSync} from 'fs'

const fn = async ()=>{
  const content = readFileSync('/math.json')

  console.log('plain content',content)
  console.log('JSON parse',JSON.parse(content))
}
  
export default await fn()
`)

console.log(result)
```

will console output this (as expected):

```bash
plain content CORRUPT THE JSON {
  "func": "async function(args){ const math = (await import('https://esm.sh/mathjs')); const result = math.evaluate(args.join(' ')); return result ;}",
  "usage": ""
}
{
  ok: false,
  error: {
    name: "SyntaxError",
    message: "unexpected token: 'CORRUPT'",
    stack: "    at <input>:1:1\n    at parse (native)\n    at fn (/src/index.js:9:3)\n    at <anonymous> (/src/index.js:12:22)\n",
  },
  isSyntaxError: true,
}
```

My assumption is, that you accidentally copy & pasted some invalid characters directly into your source code. Something which (should) never be possible to users. Otherwise, Typescript would not be able to complain about something.

In real world, you would read the file content of the JSON file, which is simply a buffer or string - no matter if the JSON itself is valid. You would not include the file content into your application source file.

**Comment by MarketingPip** [comment]
@sebastianwessel so then my stupidity. The error that is causing this is TypeScript. Can I ask you to try using in TypeScript enabled option so you can see what I am trying to explain? You will get an error trying to build.

**Comment by sebastianwessel** [comment]
I always use typescript, but I expect, that the invalid characters are removed during copy&paste here
